connector-runtime/connector-runtime-core/src/main/java/io/camunda/connector/runtime/core/secret/SecretProviderAggregator.java

```
@@ -58,6 +59,13 @@ public String getSecret(String secretName) {
     return null;
   }
 
+  @Override
+  public List<String> fetchAll(List<String> keys) {
+    return secretProviders.stream()
+        .flatMap(provider -> provider.fetchAll(keys).stream())
+        .collect(Collectors.toList());
+  }
```

Usually we only need single secret value for a key. Wouldnt this implementation possibly lead to more secret values than keys?

Or did you find a problem with the current approach?

---

>Usually we only need single secret value for a key. Wouldnt this implementation possibly lead to more secret values than keys?

Potentially yes, but the use case here is the replacement of secrets for a secure response message.

>Or did you find a problem with the current approach?

This is the motivation here: I am currently working on a custom secret provider and I found out that I must not throw exceptions from the `getSecret()` method because the replacement part is not working then. My way around this would have been to use the `fetchAll()` method to collect secrets while suppressing exceptions so that building the response message works.

>Its also important to know that we need exactly the behaviour of getSecret to make sure we get the same secrets as they were resolved before when calling getSecret.

As I saw it when doing my research, I found that you would need _at least_ the secrets you would get with `getSecret()`.

---

> but the use case here is the replacement of secrets for a secure response message.
Can you elaborate?

> I am currently working on a custom secret provider and I found out that I must not throw exceptions from the getSecret() method because the replacement part is not working then. 

The `SecretProvider`  interface describes the contract as the following:
>  @return the secret's value for the given name, if it exists. Otherwise, <code>null</code> is returned.
Why do you want to throw an exception?

It would be great if you can provide more context here.

---

>Why do you want to throw an exception?

Because I did not read and understand that properly. I would say never mind.

And thank you for tackling this together with me.

---

All good 😄